rdflib/plugins/sparql/parser.py

```diff
@@ -1235,6 +1235,15 @@ def _hexExpand(match):
     + Param("where", GroupGraphPattern),
 )
 
+# [41] WithData ::= ( 'WITH' iri ) ( DeleteData  | InsertData ) ZeroOrMore(UsingClause)
+WithData = Comp(
+    "WithData",
+    Keyword("WITH")
+    + Param("withClause", iri)
+    + (Param("delete", DeleteData) | Param("insert", InsertData))
+    + ZeroOrMore(ParamList("using", UsingClause)),
+)
+
```

Comment:
Can you clarify the rationale for this production? To me the `Modify` production is the only one that needs to accommodate `WITH`, and as far as I can tell it already does that:

https://github.com/RDFLib/rdflib/blob/c906f64b915aee28a57aae74819884fd3cab74ab/rdflib/plugins/sparql/parser.py#L1225-L1236

And it does that correctly also compared to the ratified grammar:

https://www.w3.org/TR/sparql11-query/#rModify

> ```
> [41]  	Modify	  ::=  	( 'WITH' iri )? ( DeleteClause InsertClause? | InsertClause ) UsingClause* 'WHERE' GroupGraphPattern
> ```

If there is something wrong with the `Modify` production in `rdflib/plugins/sparql/parser.py` it should be fixed instead of creating a redundant production that fixes the issue.

Reply:
The existing modify is partially complete. Ill try merging and the two and check if it works fine. Thanks for the feedback.

Reply:
> The existing modify is partially complete.

I'm actually not sure where it is incomplete, did a double check now, I may be missing something, would be good if you can point it out.